src/formatting/items.rs

```diff
@@ -1199,6 +1211,8 @@ pub(crate) fn format_trait(
         }
         let pre_block_span = if !generics.where_clause.predicates.is_empty() {
             mk_sp(generics.where_clause.span.hi(), item.span.hi())
+        } else if !generic_bounds.is_empty() {
+            mk_sp(generic_bounds.last().unwrap().span().hi(), item.span.hi())
```

Comment:
This would conflict with #4711. I'm not sure there's a real case where the generic_bounds would be empty while the collection on generics would not (or vice versa), but either way, I'd think we could just use the end of where_clause span because I _believe_ the parser advances this to desired position in the presence of any generics regardless.

Probably worth checking to confirm though.

Side note, if the span differences necessitate keeping these checks and separate calcs, that instead of checking `is_empty` and then having to call last and unwrap, an `else if let Some(last) = generic_bounds.last() {` could be used to express it a bit more succinctly

Reply:
> ... I believe the parser advances this to desired position in the presence of any generics regardless

The `item.span` is not advanced, but found that a `WhereClause` is always available, even when a where clause is not available in the code.  Therefore, removed all `if` conditions and changed to use `generics.where_clause.span.hi()` unconditionally.  (Also rebased.)

Reply:
> The item.span is not advanced, but found that a WhereClause is always available

Yes I am aware.

The spans on the AST nodes we're working with are immutable for practical purposes, and the span of the item will always cover the portions it's supposed to cover.

I was referring to the process within the parser that _sets_ those spans, and what it ends up setting as the span for the where_clause in the various cases of source input (no generics at all, generic params with no where clause, where clause with no predicates, and where clause with predicates).

I was just saying that I wasn't positive off hand where the span position of the where_clause is relative to the broader item span in those various scenarios, and that if we're going to use it unconditionally then we should validate to make sure that span will always have the hi end be placed sufficiently far enough so that we can be sure we'd avoid speciously double counting comments which _had_ actually been handled already.